web-components/src/components/form/CreditRetireForm.tsx

```diff
@@ -0,0 +1,269 @@
+import LocationStateField from '../inputs/LocationStateField';
+import ControlledTextField from '../inputs/ControlledTextField';
+import Title from '../title';
+import Description from '../description';
+import Submit from './Submit';
+import {
+  requiredMessage,
+  invalidAmount,
+  insufficientCredits,
+} from '../inputs/validation';
+
+/**
+ * This form is closely related to the form for send/transfer ecocredits (<CreditTransferForm />).
+ * In this retire form, some of its components and interfaces are exported in order to be reused in the
+ * send/transfer form, since it optionally includes the retirement of ecocredits.
+ *
+ * Retire retires a specified number of credits in the holder's account.
+ * https://docs.regen.network/modules/ecocredit/03_messages.html#msgretire
+ *
+ * Validation:
+ *    holder: must ba a valid address, and their signature must be present in the transaction
+ *    credits: must not be empty (MsgRetire.RetireCredits)
+ *      - batch_denom: must be a valid batch denomination
+ *      - amount: must be positive (aka retiredAmount)
+ *    location: must be a valid location
+ *
+ * Also:
+ * https://docs.regen.network/modules/ecocredit/protobuf.html#msgretire
+ * https://docs.regen.network/modules/ecocredit/protobuf.html#msgretire-retirecredits
+ */
+
+const useStyles = makeStyles((theme: Theme) => ({
+  groupTitle: {
+    marginTop: theme.spacing(10.75),
+    marginBottom: theme.spacing(3),
+  },
+  description: {
+    marginBottom: 0,
+    '& a': {
+      cursor: 'pointer',
+    },
+    [theme.breakpoints.up('sm')]: {
+      fontSize: theme.typography.pxToRem(16),
+    },
+    [theme.breakpoints.down('sm')]: {
+      fontSize: theme.typography.pxToRem(14),
+    },
+  },
+  textField: {
+    '& .MuiInputBase-formControl': {
+      marginTop: theme.spacing(2.25),
+    },
+  },
+  noteTextField: {
+    '& .MuiInputBase-formControl': {
+      marginTop: theme.spacing(2.25),
+    },
+    '& label': {
+      whiteSpace: 'unset',
+    },
+  },
+  stateCountryGrid: {
+    [theme.breakpoints.up('sm')]: {
+      flexWrap: 'nowrap',
+    },
+  },
+  stateCountryTextField: {
+    marginTop: theme.spacing(8),
+    [theme.breakpoints.up('sm')]: {
+      '&:first-of-type': {
+        marginRight: theme.spacing(2.375),
+      },
+      '&:last-of-type': {
+        marginLeft: theme.spacing(2.375),
+      },
+    },
+  },
+  postalCodeField: {
+    marginTop: theme.spacing(8),
+    marginBottom: theme.spacing(12.5),
+  },
+}));
+
+// Output (submit)
+interface RetireCredits {
+  batchDenom: string;
+  amount: string; // aka. retiredAmount
+}
+
+interface MsgRetire {
+  holder: string;
+  credits: RetireCredits;
+  location: string;
+  // TODO note (aka. memoNote)
+}
+
+// Input (args)
+interface FormProps {
+  holder: string;
+  batchDenom: string;
+  availableTradableAmount: number;
+  onClose: () => void;
+}
+
+export interface FormValues {
+  retiredAmount: number;
+  note: string;
+  country: string;
+  stateCountry?: string;
```

Comment:
in the field, `stateProvince` is being used as field name

Reply:
ups, thanks